src/proof/InterpolationContext.cc

```diff
@@ -9,14 +9,832 @@
+    vec<PTRef> restricted_clause;
+
+    for (Lit l : cl) {
+        Var v = var(l);
+        icolor_t var_class = getVarClass(v);
+        assert(var_class == icolor_t::I_B || var_class == icolor_t::I_A || var_class == icolor_t::I_AB);
+        icolor_t var_color = var_class == icolor_t::I_B || var_class == icolor_t::I_A ? var_class
+                                                                                      : getSharedVarColorInNode(v, n);
+        if (var_color == icolor_t::I_A || var_color == icolor_t::I_AB) {
+            PTRef ptaux = varToPTRef(v);
+            if (sign(l)) {
+                ptaux = logic.mkNot(ptaux);
+            }
+            restricted_clause.push(ptaux);
+        }
+    }
+
+    PTRef cl_ptref = logic.mkNot(logic.mkOr(std::move(restricted_clause)));
+    PTRef implicant = logic.mkAnd({pmanager.getPartition(A_mask, PartitionManager::part::A), cl_ptref});
+
+    bool res = VerificationUtils(config, logic).impliesExternal(implicant, getPartialInterpolant(n));
+    assert(res);
+    return res;
+}
+
+bool SingleInterpolationComputationContext::verifyPartialInterpolantB(ProofNode const & n) {
+    // Check B /\ ~(C|b,ab) -> ~I, i.e., B /\ ~(C|b,ab) /\ I unsat
+    std::vector<Lit> const & cl = n.getClause();
+    vec<PTRef> restricted_clause;
+
+    for (Lit l : cl) {
+        Var v = var(l);
+        icolor_t var_class = getVarClass(v);
+        assert(var_class == icolor_t::I_B || var_class == icolor_t::I_A || var_class == icolor_t::I_AB);
+        icolor_t var_color = var_class == icolor_t::I_B || var_class == icolor_t::I_A ? var_class
+                                                                                      : getSharedVarColorInNode(v, n);
+        if (var_color == icolor_t::I_B || var_color == icolor_t::I_AB) {
+            PTRef ptaux = varToPTRef(v);
+            if (sign(l)) {
+                ptaux = logic.mkNot(ptaux);
+            }
+            restricted_clause.push(ptaux);
+        }
+    }
+
+    PTRef cl_ptref = logic.mkNot(logic.mkOr(std::move(restricted_clause)));
+    PTRef implicant = logic.mkAnd(pmanager.getPartition(A_mask, PartitionManager::part::B), cl_ptref);
+
+    bool res = VerificationUtils(config, logic).impliesExternal(implicant, logic.mkNot(getPartialInterpolant(n)));
+    assert(res);
+    return res;
+}
+
+
+SingleInterpolationComputationContext::SingleInterpolationComputationContext(
+        const SMTConfig & config,
+        Theory & theory,
+        TermMapper & termMapper,
+        PartitionManager & pmanager,
+        const ProofGraph & proof,
+        const ipartitions_t & A_mask
+) : logic(theory.getLogic()), config(config), pmanager(pmanager), proofGraph(proof), thandler(new THandler(theory, termMapper)), A_mask(A_mask) {
+    auto const & vars = proof.getVariables();
+    std::size_t varCounts = (*std::max_element(vars.begin(), vars.end())) + 1;
+    nodeData.clear();
+    AB_vars_mapping.clear();
+    nodeData.resize(proof.getGraphSize());
```

Comment:
It looks like the number of nodes in the proof might be exaggerated by this, since the proof stores the nodes in a vector and during reduction, some nodes are removed, but they leave a hole in this vector behind.
Compacting the vector seems to be problematic, since the nodes themselves remember their index as id, and this id is used in multiple other places.

The effect is that the interpolation data might be bigger than necessary (prepared for more nodes). It is not clear how big the effect on memory consumption is. It might be negligible.
We could store pointers to `InterpolationNodeData` instead of the objects themselves, which would mimic the old setup. But that also has some memory overhead (compared to the current version) for the nodes that are truly present in the proof.
Currently, I am inclined to keep the current version.

Reply:
I'm ok with going with your suggestion.  If it turns out in later measurements that it's a performance issue, or something that can be improved on, we can do it at that point.